File: src/0721_accounts_merge.js
class UnionFind {
    constructor(n) {
        this.parent = Array.from({ length: n }, (_, i) => i);
        this.rank = Array(n).fill(0);
    }
    find(x) {
        while (this.parent[x] !== x) {
            this.parent[x] = this.parent[this.parent[x]];
            x = this.parent[x];
        }
        return x;
    }
    union(a, b) {
        const pa = this.find(a);
        const pb = this.find(b);
        if (pa === pb) return;
        if (this.rank[pa] < this.rank[pb]) this.parent[pa] = pb;
        else if (this.rank[pa] > this.rank[pb]) this.parent[pb] = pa;
        else {
            this.parent[pb] = pa;
            this.rank[pa]++;
        }
    }
}

function accountsMerge(accounts) {
    const uf = new UnionFind(accounts.length);
    const emailToAcc = new Map();
    for (let i = 0; i < accounts.length; i++) {
        for (let j = 1; j < accounts[i].length; j++) {
            const email = accounts[i][j];
            if (emailToAcc.has(email)) uf.union(i, emailToAcc.get(email));
            else emailToAcc.set(email, i);
        }
    }
    const emailGroup = new Map();
    for (const [email, i] of emailToAcc.entries()) {
        const leader = uf.find(i);
        if (!emailGroup.has(leader)) emailGroup.set(leader, []);
        emailGroup.get(leader).push(email);
    }
    const res = [];
    for (const [i, emails] of emailGroup.entries()) {
        res.push([accounts[i][0], ...emails.sort()]);
    }
    return res;
}

module.exports = { accountsMerge };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(accountsMerge([["John", "johnsmith@mail.com", "john_newyork@mail.com"], ["John", "johnsmith@mail.com", "john00@mail.com"], ["Mary", "mary@mail.com"], ["John", "johnnybravo@mail.com"]]), [["John", "john00@mail.com", "john_newyork@mail.com", "johnsmith@mail.com"], ["Mary", "mary@mail.com"], ["John", "johnnybravo@mail.com"]]);
    assert.deepStrictEqual(accountsMerge([["Gabe", "Gabe0@m.co", "Gabe3@m.co", "Gabe1@m.co"], ["Kevin", "Kevin3@m.co", "Kevin5@m.co", "Kevin0@m.co"], ["Ethan", "Ethan5@m.co", "Ethan4@m.co", "Ethan0@m.co"], ["Hanzo", "Hanzo3@m.co", "Hanzo1@m.co", "Hanzo0@m.co"], ["Fern", "Fern5@m.co", "Fern1@m.co", "Fern0@m.co"]]), [["Gabe", "Gabe0@m.co", "Gabe1@m.co", "Gabe3@m.co"], ["Kevin", "Kevin0@m.co", "Kevin3@m.co", "Kevin5@m.co"], ["Ethan", "Ethan0@m.co", "Ethan4@m.co", "Ethan5@m.co"], ["Hanzo", "Hanzo0@m.co", "Hanzo1@m.co", "Hanzo3@m.co"], ["Fern", "Fern0@m.co", "Fern1@m.co", "Fern5@m.co"]]);
}

File: src/0721_accounts_merge.rs
struct Solution;

use std::collections::HashMap;

struct UnionFind {
    parent: Vec<usize>,
    rank: Vec<i32>,
}

impl UnionFind {
    fn new(size: usize) -> Self {
        Self {
            parent: (0..size).collect(),
            rank: vec![1; size],
        }
    }

    fn find(&mut self, node: usize) -> usize {
        if self.parent[node] != node {
            let root = self.find(self.parent[node]);
            self.parent[node] = root;
        }
        self.parent[node]
    }

    fn union(&mut self, a: usize, b: usize) {
        let root_a = self.find(a);
        let root_b = self.find(b);
        if root_a == root_b {
            return;
        }
        if self.rank[root_a] < self.rank[root_b] {
            self.parent[root_a] = root_b;
        } else if self.rank[root_a] > self.rank[root_b] {
            self.parent[root_b] = root_a;
        } else {
            self.parent[root_b] = root_a;
            self.rank[root_a] += 1;
        }
    }
}

impl Solution {
    pub fn accounts_merge(accounts: Vec<Vec<String>>) -> Vec<Vec<String>> {
        let mut uf = UnionFind::new(accounts.len());
        let mut email_to_acc: HashMap<String, usize> = HashMap::new();
        let mut email_order: Vec<(String, usize)> = Vec::new();

        for (i, account) in accounts.iter().enumerate() {
            for email in account.iter().skip(1) {
                if let Some(&prev) = email_to_acc.get(email) {
                    uf.union(i, prev);
                } else {
                    email_to_acc.insert(email.clone(), i);
                    email_order.push((email.clone(), i));
                }
            }
        }

        let mut email_group: Vec<(usize, Vec<String>)> = Vec::new();
        for (email, i) in email_order {
            let leader = uf.find(i);
            if let Some((_, emails)) = email_group.iter_mut().find(|(idx, _)| *idx == leader) {
                emails.push(email);
            } else {
                email_group.push((leader, vec![email]));
            }
        }

        let mut res = Vec::new();
        for (i, mut emails) in email_group {
            emails.sort();
            let mut row = vec![accounts[i][0].clone()];
            row.extend(emails);
            res.push(row);
        }
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn mirrors_python_main_assertions() {
        assert_eq!(
            Solution::accounts_merge(vec![
                vec![
                    "John".to_string(),
                    "johnsmith@mail.com".to_string(),
                    "john_newyork@mail.com".to_string()
                ],
                vec![
                    "John".to_string(),
                    "johnsmith@mail.com".to_string(),
                    "john00@mail.com".to_string()
                ],
                vec!["Mary".to_string(), "mary@mail.com".to_string()],
                vec!["John".to_string(), "johnnybravo@mail.com".to_string()]
            ]),
            vec![
                vec![
                    "John".to_string(),
                    "john00@mail.com".to_string(),
                    "john_newyork@mail.com".to_string(),
                    "johnsmith@mail.com".to_string()
                ],
                vec!["Mary".to_string(), "mary@mail.com".to_string()],
                vec!["John".to_string(), "johnnybravo@mail.com".to_string()]
            ]
        );
        assert_eq!(
            Solution::accounts_merge(vec![
                vec![
                    "Gabe".to_string(),
                    "Gabe0@m.co".to_string(),
                    "Gabe3@m.co".to_string(),
                    "Gabe1@m.co".to_string()
                ],
                vec![
                    "Kevin".to_string(),
                    "Kevin3@m.co".to_string(),
                    "Kevin5@m.co".to_string(),
                    "Kevin0@m.co".to_string()
                ],
                vec![
                    "Ethan".to_string(),
                    "Ethan5@m.co".to_string(),
                    "Ethan4@m.co".to_string(),
                    "Ethan0@m.co".to_string()
                ],
                vec![
                    "Hanzo".to_string(),
                    "Hanzo3@m.co".to_string(),
                    "Hanzo1@m.co".to_string(),
                    "Hanzo0@m.co".to_string()
                ],
                vec![
                    "Fern".to_string(),
                    "Fern5@m.co".to_string(),
                    "Fern1@m.co".to_string(),
                    "Fern0@m.co".to_string()
                ]
            ]),
            vec![
                vec![
                    "Gabe".to_string(),
                    "Gabe0@m.co".to_string(),
                    "Gabe1@m.co".to_string(),
                    "Gabe3@m.co".to_string()
                ],
                vec![
                    "Kevin".to_string(),
                    "Kevin0@m.co".to_string(),
                    "Kevin3@m.co".to_string(),
                    "Kevin5@m.co".to_string()
                ],
                vec![
                    "Ethan".to_string(),
                    "Ethan0@m.co".to_string(),
                    "Ethan4@m.co".to_string(),
                    "Ethan5@m.co".to_string()
                ],
                vec![
                    "Hanzo".to_string(),
                    "Hanzo0@m.co".to_string(),
                    "Hanzo1@m.co".to_string(),
                    "Hanzo3@m.co".to_string()
                ],
                vec![
                    "Fern".to_string(),
                    "Fern0@m.co".to_string(),
                    "Fern1@m.co".to_string(),
                    "Fern5@m.co".to_string()
                ]
            ]
        );
    }
}
